Add a `pyproject.toml` and minimal packaging structure

FollowTheProcess opened the pull request:
Hey! 👋🏻

Saw this project come up on my GitHub feed and figured it's quick enough for me to help out with some of the issues you've put up.

This PR closes #5

I've put a fairly minimal `pyproject.toml` in using hatch. I've had to restructure the package slightly to make this work (I'm not sure if single file packages are possible, at least I've never done one). I've gone for `./src` layout but let me know if you'd like a different layout!

Also added `pytest` as a dev dependency with a config I've found useful in my projects. And added an entry point 🙂

Let me know if you'd like any changes or have any other comments 👍🏻

brettcannon commented on pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+[project.urls]
+source = "https://github.com/brettcannon/microvenv"
+
+[project.scripts]
+microvenv = "microvenv.main:main"
```
I [don't want an entry point](https://github.com/brettcannon/microvenv/issues/5#issuecomment-1455219067).

FollowTheProcess replied:
Removed :)

brettcannon commented on pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+[project.scripts]
+microvenv = "microvenv.main:main"
+
+[tool.pytest.ini_options]
+minversion = "7.0"
+addopts = [ "-ra", "--strict-markers", "--strict-config" ]
+xfail_strict = true
+log_cli_level = "info"
+pythonpath = [ "src" ]
+testpaths = [ "tests" ]
```
What is all of this? Testing has not been set up yet.

FollowTheProcess replied:
Just some helpful pytest config I copied from one of my projects. Agree though it doesn't make much sense here as the project is so simple

brettcannon reviewed (changes requested):
The project structure is unfortunately very purposeful, so I can't have this converted into a package; it needs to stay a self-contained file. As such, I don't think having a `src` layout is helpful and just makes running the tests a little bit harder by requiring an editable install for not benefit.

brettcannon commented:
Thanks for the PR!

> I've put a fairly minimal `pyproject.toml` in using hatch. I've had to restructure the package slightly to make this work (I'm not sure if single file packages are possible, at least I've never done one). I've gone for `./src` layout but let me know if you'd like a different layout!

Unfortunately I need this to stay a single file.



> And added an entry point 🙂

Unfortunately I don't want an entry point (see the original issue as to why).

brettcannon commented:
And FYI [Flit](https://flit.pypa.io/) supports a single file if Hatch doesn't.

FollowTheProcess commented on pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+  "Development Status :: 3 - Alpha",
+  "Environment :: Console",
+  "Intended Audience :: Developers",
+  "License :: OSI Approved :: Apache Software License",
```
Copy paste strikes again 🤦🏻‍♂️

FollowTheProcess replied:
Fixed

FollowTheProcess commented on pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+  "Operating System :: POSIX :: Linux",
+  "Operating System :: Microsoft :: Windows",
+  "Operating System :: OS Independent",
+  "Programming Language :: Python :: 3",
+  "Programming Language :: Python :: 3 :: Only",
+  "Programming Language :: Python :: 3.7",
+  "Programming Language :: Python :: 3.8",
+  "Programming Language :: Python :: 3.9",
+  "Programming Language :: Python :: 3.10",
+  "Programming Language :: Python :: 3.11",
```
I've kept this in as I've seen you've now got CI on all these set up

FollowTheProcess commented on pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+  "Operating System :: MacOS :: MacOS X",
+  "Operating System :: POSIX :: Linux",
+  "Operating System :: Microsoft :: Windows",
+  "Operating System :: OS Independent",
```
Agreed, more copy paste. Got rid and just kept OS Independent 👍🏻

FollowTheProcess commented:
Looks like hatch supports a single file package out of the box, I did `hatch build` and unzipped the tarball it looks as if everything is there as expected.

Also noticed you've now added ruff as a linter. I have a fairly strict `ruff.toml` in my home directory that was making it trigger a load of warnings. I think just including the default ruff config in `pyproject.toml` should stop that happening for anyone else

Think I addressed all your comments, let me know if you want me to make any further changes 👍🏻

brettcannon commented:
I marked all of the requested changes as resolved and refreshed my review. Unfortunately it looks like the changes you made to `microvenv.py` aren't formatted with Black. But since this is about to become a blocker for me releasing this project to PyPI, I'm going to override the failure and fix the code up myself.

Thanks for the PR!
